src/Middleware/OutputCaching/src/Memory/MemoryOutputCacheStore.cs

```diff
@@ -26,12 +31,29 @@ public ValueTask EvictByTagAsync(string tag, CancellationToken cancellationToken
         {
             if (_taggedEntries.TryGetValue(tag, out var keys))
             {
-                foreach (var key in keys)
+                if (keys != null && keys.Count > 0)
                 {
-                    _cache.Remove(key);
-                }
+                    // Since eviction callbacks run inline, iterating over keys could throw
```

Comment:
nit:
```suggestion
                    // Since eviction callbacks can potentially run inline, iterating over keys could throw
```

Reply:
Good catch! I'd go even further and say:

```suggestion
                    // If MemoryCache changed to run eviction callbacks run inline in Remove, iterating over keys could throw
```

And I don't know why `MemoryCache` would start evictions inline in `Remove` of all places, but I guess it's possible.

Also, can we change the type of `_cache` from `IMemoryCache` to `MemoryCache`? That's the only thing we call the internal constructor with, right? If we want to make it more abstract later on, we can. But for now, let's be clear that this is tightly integrated.

Reply:
I actually liked it being `IMemoryCache`: https://github.com/dotnet/aspnetcore/pull/43728#discussion_r965245186 😄 
But looks like `MemoryCache` isn't sealed so maybe the same thing can be done.

Also, editing your suggestion:
```suggestion
                    // If MemoryCache changed to run eviction callbacks inline in Remove, iterating over keys could throw
```